transaction/ResourceId.cpp

```diff
@@ -0,0 +1,99 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#include "transaction/ResourceId.hpp"
+
+#include <cstddef>
+#include <string>
+#include <utility>
+
+namespace quickstep {
+
+namespace transaction {
+
+bool ResourceId::operator==(const ResourceId &other) const {
+  return (db_id_ == other.db_id_) &&
+    (rel_id_ == other.rel_id_) &&
+    (block_id_ == other.block_id_) &&
+    (tuple_id_ == other.tuple_id_);
+}
+
+bool ResourceId::hasParent() const {
+  return !isDatabaseAccess();
+}
+
+ResourceId ResourceId::getParentResourceId() const {
+  if (isRelationAccess()) {
+    return ResourceId(db_id_);
+  } else if (isBlockAccess()) {
+    return ResourceId(db_id_, rel_id_);
+  } else if (isTupleAccess()) {
+    return ResourceId(db_id_, rel_id_, block_id_);
+  } else {
+    FATAL_ERROR("Trying to call getParentResourceId on database");
+  }
+}
+
+bool ResourceId::isDatabaseAccess() const {
+  return (db_id_ != kDbIdPlaceholder) &&
+    (rel_id_ == kRelIdPlaceholder) &&
+    (block_id_ == kBlockIdPlaceholder) &&
+    (tuple_id_ == kTupleIdPlaceholder);
```

Comment:
Actually, I suggest to provide four `private` methods to check whether the given four resource id is the placeholder, and then use the combinations of the four methods for this method and three more below.